File: scripts/fetch-product-feed.py
#!/usr/bin/env python3

"""
Fetches the Shopify Atom feed and converts it to markdown for Pinecone upload
Run: python3 scripts/fetch-product-feed.py
"""

import xml.etree.ElementTree as ET
import urllib.request
import re
from datetime import date

FEED_URL = 'https://portablespas.co.nz/collections/all.atom'
OUTPUT_FILE = 'product-catalog.md'

# XML namespaces
NAMESPACES = {
    'atom': 'http://www.w3.org/2005/Atom',
    's': 'http://jadedpixel.com/-/spec/shopify'
}

def fetch_feed():
    """Fetch the Atom feed from Shopify"""
    print('Fetching product feed from Shopify...')
    with urllib.request.urlopen(FEED_URL) as response:
        return response.read().decode('utf-8')

def strip_html(text):
    """Remove HTML tags and clean up text"""
    if not text:
        return ''
    # Remove CDATA
    text = re.sub(r'<!\[CDATA\[(.*?)\]\]>', r'\1', text, flags=re.DOTALL)
    # Remove HTML tags
    text = re.sub(r'<[^>]+>', '', text)
    # Clean up entities
    text = text.replace('&nbsp;', ' ')
    text = text.replace('&amp;', '&')
    text = text.replace('&lt;', '<')
    text = text.replace('&gt;', '>')
    text = text.replace('&quot;', '"')
    # Clean up whitespace
    text = re.sub(r'\s+', ' ', text)
    return text.strip()

def extract_description(summary):
    """Extract clean description from summary HTML"""
    if not summary:
        return ''

    # Try to extract text from the second table cell (description area)
    match = re.search(r'<td colspan="2">(.*?)</td>', summary, re.DOTALL)
    if match:
        desc = strip_html(match.group(1))
        # Remove "Vendor:", "Type:", "Price:" lines if present
        desc = re.sub(r'(Vendor:|Type:|Price:).*?(\n|$)', '', desc)
        return desc.strip()

    return strip_html(summary)

def parse_feed(xml_content):
    """Parse the Atom feed and extract product information"""
    print('Parsing feed...')
    root = ET.fromstring(xml_content)

    products = []

    for entry in root.findall('atom:entry', NAMESPACES):
        title = entry.find('atom:title', NAMESPACES)
        link = entry.find('atom:link[@rel="alternate"]', NAMESPACES)
        summary = entry.find('atom:summary', NAMESPACES)
        product_type = entry.find('s:type', NAMESPACES)
        vendor = entry.find('s:vendor', NAMESPACES)

        # Get variant info (price)
        variant = entry.find('s:variant', NAMESPACES)
        price = None
        sku = None
        if variant is not None:
            price_elem = variant.find('s:price', NAMESPACES)
            sku_elem = variant.find('s:sku', NAMESPACES)
            if price_elem is not None:
                price = price_elem.text
            if sku_elem is not None and sku_elem.text:
                sku = sku_elem.text

        # Get tags
        tags = [tag.text for tag in entry.findall('s:tag', NAMESPACES)]

        product = {
            'title': title.text if title is not None else 'Unknown',
            'link': link.get('href') if link is not None else '',
            'price': price,
            'sku': sku,
            'type': product_type.text if product_type is not None else 'Other',
            'vendor': vendor.text if vendor is not None else '',
            'description': extract_description(summary.text if summary is not None else ''),
            'tags': tags
        }

        products.append(product)

    return products

def convert_to_markdown(products):
    """Convert products to markdown format"""
    print('Converting to markdown...')

    md = f"# Portable Spas Product Catalog\n\n"
    md += f"*Last updated: {date.today().isoformat()}*\n\n"
    md += f"This document contains the current product catalog for Portable Spas New Zealand.\n\n"
    md += f"Total products: {len(products)}\n\n"
    md += f"---\n\n"

    # Group by type
    by_type = {}
    for product in products:
        ptype = product['type']
        if ptype not in by_type:
            by_type[ptype] = []
        by_type[ptype].append(product)

    # Sort types
    for ptype in sorted(by_type.keys()):
        md += f"## {ptype}\n\n"

        for product in by_type[ptype]:
            md += f"### {product['title']}\n\n"

            if product['price']:
                md += f"**Price:** ${product['price']} NZD\n\n"

            md += f"**Product Page:** {product['link']}\n\n"

            if product['sku']:
                md += f"**SKU:** {product['sku']}\n\n"

            if product['tags']:
                md += f"**Tags:** {', '.join(product['tags'])}\n\n"

            if product['description']:
                md += f"{product['description']}\n\n"

            md += f"---\n\n"

    return md

def main():
    try:
        # Fetch and parse
        xml_content = fetch_feed()
        products = parse_feed(xml_content)

        print(f"Found {len(products)} products")

        # Convert to markdown
        markdown = convert_to_markdown(products)

        # Write to file
        print(f'Writing to {OUTPUT_FILE}...')
        with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
            f.write(markdown)

        print('✅ Done! Product catalog saved to product-catalog.md')
        print('\nYou can now upload this file to Pinecone via the admin interface at:')
        print('https://portable-spas-ai-chat.vercel.app/admin')

    except Exception as e:
        print(f'❌ Error: {e}')
        import traceback
        traceback.print_exc()
        exit(1)

if __name__ == '__main__':
    main()

File: scripts/fetch-product-feed.js
#!/usr/bin/env node

/**
 * Fetches the Shopify Atom feed and converts it to markdown for Pinecone upload
 * Run: node scripts/fetch-product-feed.js
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

const FEED_URL = 'https://portablespas.co.nz/collections/all.atom';
const OUTPUT_FILE = path.join(__dirname, '..', 'product-catalog.md');

function fetchFeed() {
  return new Promise((resolve, reject) => {
    https.get(FEED_URL, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        resolve(data);
      });
    }).on('error', (err) => {
      reject(err);
    });
  });
}

function parseAtomFeed(xml) {
  const products = [];

  // Simple XML parsing - extract entry elements
  const entryRegex = /<entry>(.*?)<\/entry>/gs;
  const entries = xml.match(entryRegex) || [];

  entries.forEach(entry => {
    // Extract fields
    const title = (entry.match(/<title>(.*?)<\/title>/s) || [])[1]?.trim();
    const link = (entry.match(/<link[^>]*href="([^"]*)"/) || [])[1];
    const summary = (entry.match(/<summary[^>]*>(.*?)<\/summary>/s) || [])[1];
    const productType = (entry.match(/<s:type>(.*?)<\/s:type>/) || [])[1];
    const vendor = (entry.match(/<s:vendor>(.*?)<\/s:vendor>/) || [])[1];

    // Extract price
    const priceMatch = entry.match(/\$([0-9,]+\.[0-9]{2})/);
    const price = priceMatch ? priceMatch[0] : null;

    // Extract description from summary (remove HTML tags)
    let description = '';
    if (summary) {
      description = summary
        .replace(/<[^>]*>/g, '') // Remove HTML tags
        .replace(/&nbsp;/g, ' ')
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&quot;/g, '"')
        .trim();
    }

    if (title && link) {
      products.push({
        title,
        link,
        price,
        productType,
        vendor,
        description
      });
    }
  });

  return products;
}

function convertToMarkdown(products) {
  let markdown = `# Portable Spas Product Catalog\n\n`;
  markdown += `*Last updated: ${new Date().toISOString().split('T')[0]}*\n\n`;
  markdown += `This document contains the current product catalog for Portable Spas New Zealand.\n\n`;
  markdown += `---\n\n`;

  // Group by product type
  const grouped = {};
  products.forEach(product => {
    const type = product.productType || 'Other';
    if (!grouped[type]) {
      grouped[type] = [];
    }
    grouped[type].push(product);
  });

  // Output by category
  Object.keys(grouped).sort().forEach(type => {
    markdown += `## ${type}\n\n`;

    grouped[type].forEach(product => {
      markdown += `### ${product.title}\n\n`;

      if (product.price) {
        markdown += `**Price:** ${product.price} NZD\n\n`;
      }

      markdown += `**Product Page:** ${product.link}\n\n`;

      if (product.description) {
        markdown += `**Description:** ${product.description}\n\n`;
      }

      markdown += `---\n\n`;
    });
  });

  return markdown;
}

async function main() {
  try {
    console.log('Fetching product feed from Shopify...');
    const xml = await fetchFeed();

    console.log('Parsing feed...');
    const products = parseAtomFeed(xml);

    console.log(`Found ${products.length} products`);

    console.log('Converting to markdown...');
    const markdown = convertToMarkdown(products);

    console.log(`Writing to ${OUTPUT_FILE}...`);
    fs.writeFileSync(OUTPUT_FILE, markdown, 'utf8');

    console.log('✅ Done! Product catalog saved to product-catalog.md');
    console.log('\nYou can now upload this file to Pinecone via the admin interface at:');
    console.log('https://portable-spas-ai-chat.vercel.app/admin');
  } catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
}

main();
